### PR TITLE
Added wander/wanderUnordered to Task.

### PR DESCRIPTION
**Pull request by mutantmell**
This should satisfy Issue #249.

I ended up modifying the backing implementation of `TaskGather` to requiring multiple CanBuildFroms for gatherUnordered.

I added a first pass at the Scaladoc, based heavily on the Scaladoc for gather(unordered).

Let me know if there's any questions/comments/concerns.

-Matt

### CODE REVIEW

File: monix-eval/shared/src/main/scala/monix/eval/Task.scala [hunk 1]
```diff
@@ -875,7 +875,29 @@ object Task extends TaskInstances {
     */
   def gather[A, M[X] <: TraversableOnce[X]](in: M[Task[A]])
     (implicit cbf: CanBuildFrom[M[Task[A]], A, M[A]]): Task[M[A]] =
-    TaskGather(in)(cbf)
+    TaskGather(in.toArray, () => cbf(in))
```
Comment (alexandru) [inline monix-eval/shared/src/main/scala/monix/eval/Task.scala]: Problem with this `.toArray` is that depending on the implementation of `M[A]`:

1. this code can throw exceptions, leading to a different result than the `TaskGather` that took `in` directly (see that the original `.toArray` was inside that `try/catch`)
2. a `.toArray` on a large data-structure or on some Java Iterable that blocks threads can be an expensive operation (I know it sounds awful, but there are some truly awful Java implementations out there) and by not processing this in the context of the task, the user can get no opportunity to process it on a different thread

It's also why the second parameter is a `Function1`. My suggestion is to make `TaskGather` take a `TraversableOnce[Task[A]]` parameter and move the `.toArray` inside `TaskGather` again. Or maybe make this parameter lazy as well?

File: monix-eval/shared/src/main/scala/monix/eval/Task.scala [hunk 2]
```diff
@@ -875,7 +875,29 @@ object Task extends TaskInstances {
...
+   *
+   * Although the effects are unordered, we ensure the order of results
+   * matches the order of the input sequence. Also see [[wanderUnordered]]
+   * for the more efficient alternative.
+   *
+   * It's a generalized version of [[gather]].
+   */
+  def wander[A, B, M[X] <: TraversableOnce[X]](in: M[A])(f: A => Task[B])
+    (implicit cbf: CanBuildFrom[M[A], B, M[B]]): Task[M[B]] = {
+    TaskGather(in.map(f).toArray, () => cbf(in))
```
Comment (alexandru) [inline monix-eval/shared/src/main/scala/monix/eval/Task.scala]: I would have done it like ... 

```scala
def wander[A, B, M[X] <: TraversableOnce[X]](in: M[A])(f: A => Task[B])
  (implicit cbf: CanBuildFrom[M[A], A, M[A]]): Task[M[B]] = {

  Task.eval(in.map(f)).flatMap(col => TaskGather(col, () => cbf(in)))
}
```

This way `in.map(f)` is protected from thrown exceptions and you don't have to turn it into an array.

File: monix-eval/shared/src/main/scala/monix/eval/Task.scala [hunk 3]
```diff
@@ -895,6 +917,30 @@ object Task extends TaskInstances {
...
+    * results will be ordered. Useful when you don't need ordering because:
+    *
+    *  - it has non-blocking behavior (but not wait-free)
+    *  - it can be more efficient (compared with [[wander]]), but not
+    *    necessarily (if you care about performance, then test)
+    *
+    * It's a generalized version of [[gatherUnordered]].
+    */
+  def wanderUnordered[A, B, M[X] <: TraversableOnce[X]](in: M[A])(f: A => Task[B])
+    (implicit cbf: CanBuildFrom[M[A], B, M[B]]): Task[M[B]] = {
```
Comment (alexandru) [inline monix-eval/shared/src/main/scala/monix/eval/Task.scala]: I would not go with a version with `CanBuildFrom` and simply follow the normal `gatherUnordered` signature.

This is new in master and the version 2.1 I'm going to release (probably today), but the reason is that we are no longer using `CanBuildFrom` in `gatherUnordered` and returning a `Task[List[A]]` instead is because it better reflects what the function does, because the result really is a `List`, being kept in an `Atomic` and updated in a non-blocking fashion and if the user wants something else, he can convert that `List` himself, with whatever cost that incurs and that cost will be visible, whereas if you hide it in this function, that cost will not be visible.

So basically traversing that `List` to build the collection you want is not worth it imho. It's also the defining characteristic versus normal `gather`. I mean originally `gather` was built on top of `gatherUnordered`, following Scalaz in this regard, but now `gather` has a pretty efficient version too, so the defining characteristics of `gatherUnordered` are:

1. non-blocking
2. more memory efficient, since we traverse the original collection only once and build the final `List` as we go along, only once

Let me know your thoughts on this matter.

File: monix-eval/shared/src/main/scala/monix/eval/Task.scala (hunk 1 above)
Comment (mutantmell) [inline monix-eval/shared/src/main/scala/monix/eval/Task.scala]: I changed `TaskGather` to take a `TraversableOnce[Task[A]]` parameter - this helped address the next issue as well.

(the `.toArray` being there is an artifact of a previous draft, when the code still took in `CanBuildFrom` instead of a builder maker)

File: monix-eval/shared/src/main/scala/monix/eval/Task.scala (hunk 2 above)
Comment (mutantmell) [inline monix-eval/shared/src/main/scala/monix/eval/Task.scala]: Done

File: monix-eval/shared/src/main/scala/monix/eval/Task.scala (hunk 3 above)
Comment (mutantmell) [inline monix-eval/shared/src/main/scala/monix/eval/Task.scala]: No objections here

Comment (mutantmell) [inline monix-eval/shared/src/main/scala/monix/eval/Task.scala]: Im adding tests to verify these properties, then I'll update the pr

**Comment by alexandru** [comment]
Thanks @mutantmell - version 2.1 is already released on Maven Central, your PR will make it in the next release.
